Split SMS/app flows for two-factor authentication setup

xknown opened the pull request:
#### Changes proposed in this Pull Request

Removes the requirement to enter a phone number when setting up two-factor authentication.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Log in to a WP.com account without 2fa enabled
* Navigate to /me/security/two-step
* Activate 2fa with an app, or SMS. Make sure it works as usual.

##### Before:

 1. Initial screen <img width="954" alt="Screenshot 2021-10-28 at 20 58 17" src="https://user-images.githubusercontent.com/1081870/139318781-98294766-3d19-4dcd-a0c9-2c9f1374c8c9.png">
 2. Setup screen SMS/App <img width="965" alt="Screenshot 2021-10-28 at 20 58 38" src="https://user-images.githubusercontent.com/1081870/139318789-2236627a-731f-468d-a4f4-f35153b1e506.png">

##### After:
 1. Initial screen <img width="971" alt="Screenshot 2021-10-28 at 20 44 21" src="https://user-images.githubusercontent.com/1081870/139318955-82b4303d-da2e-4786-8360-ddde014c9951.png">
 2. Setup with an app
    * QR code <img width="965" alt="Screenshot 2021-10-28 at 20 44 56" src="https://user-images.githubusercontent.com/1081870/139318957-2c1ed36a-8313-445e-987d-4f22b09c1dc7.png">
    * Backup codes <img width="967" alt="Screenshot 2021-10-28 at 20 51 27" src="https://user-images.githubusercontent.com/1081870/139318959-5ded8c25-72e5-415d-a98f-e940b2b8d449.png">
 3. Setup with SMS
    * Phone number <img width="974" alt="Screenshot 2021-10-28 at 20 51 57" src="https://user-images.githubusercontent.com/1081870/139318963-4360303d-1687-4b81-83cc-58e5340caec1.png">
    * Code verification <img width="964" alt="Screenshot 2021-10-28 at 20 52 18" src="https://user-images.githubusercontent.com/1081870/139318964-3ffc942f-dc4c-4de2-be42-df9f14e261cb.png">
    * Backup codes <img width="973" alt="Screenshot 2021-10-28 at 20 52 39" src="https://user-images.githubusercontent.com/1081870/139318966-d7e12664-a67c-41da-bac5-ac98f50f840e.png">


<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #36063

xknown committed:
Initial bits to separate sms/app flows for 2fa setup

Fixes #36063

a8ci18n commented on client/me/security-2fa-progress/index.jsx:
```diff
@@ -31,7 +31,7 @@ class Security2faProgress extends Component {
 				<div className="security-2fa-progress__inner-container">
 					<ProgressItem
-						label={ this.props.translate( 'Enter Phone Number' ) }
+						label={ this.props.translate( 'Authentication Method' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **14** times:
`translate( 'Invalid authentication method' )` *ES Score: 7*

xknown replied:
Can't really use the suggested string.

xknown committed:
Keep the current selected auth method when one clicks the Cancel button.

xknown committed:
Show the correct steps depending on the chosen flow

xknown committed:
Use the correct variable

tyxla commented on client/me/security-2fa-setup/index.jsx:
```diff
@@ -81,7 +85,10 @@ class Security2faSetup extends Component {
 				{ 'backup-codes' === this.state.step ? (
-					<Security2faSetupBackupCodes onFinished={ this.onFinished } />
+					<Security2faSetupBackupCodes
+						doSMSFlow={ this.state.authMethod !== 'app-based' }
```
Minor consistency question: we usually name those flags `isSomeFlow` rather than `doSomeFlow`. Would it make sense if we rename the prop to `isSmsFlow`?

xknown replied:
Fixed in ffd7fd7

tyxla commented on client/me/security-2fa-progress/index.jsx:
```diff
@@ -27,14 +27,17 @@ class Security2faProgress extends Component {
 	};
 
 	render() {
+		const doSMSFlow = this.props.doSMSFlow ?? false;
```
Can't we just destructure from `this.props` here?

```suggestion
		const { doSMSFlow } = this.props;
```

tyxla commented on client/me/security-2fa-initial-setup/style.scss:
```diff
@@ -0,0 +1,11 @@
+
+.security-2fa-initial-setup__label p {
+    font-weight: normal;
+}
```
Let's add an empty new line at the end of this one to make the linter happy.

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,12 +54,39 @@ class Security2faInitialSetup extends Component {
+							name="auth_method"
+							value="app-based"
+							defaultChecked={ isAppSelected }
+							onChange={ this.setAuth.bind( this ) }
```
Instead of binding to `this`, we can just declare `setAuth` as an arrow function above. That will automatically bind it to `this`.

xknown replied:
Updated in https://github.com/Automattic/wp-calypso/pull/57434/commits/99f043564b82d9dbb141bcc16afa1f10108b9202

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,12 +54,39 @@ class Security2faInitialSetup extends Component {
+							onChange={ this.setAuth.bind( this ) }
+						/>
+						<Gridicon icon="phone" />
+						<span>Set up using an app</span>
```
Should we localize all those new strings? I can imagine we'll need to eventually translate them.

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,12 +54,39 @@ class Security2faInitialSetup extends Component {
+						<Gridicon icon="phone" />
+						<span>Set up using an app</span>
+						<p>
+							Use an application on your phone to get two-factor authentication codes when you log
```
Should we get a copy review on all those strings, or are they already approved?

xknown replied:
Who needs to approve them? :)

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,12 +54,39 @@ class Security2faInitialSetup extends Component {
-						this.props.onSuccess( event );
-					} }
-				>
+				<div>
```
Do we need that `div`?

tyxla commented on client/me/security-2fa-initial-setup/style.scss:
```diff
@@ -0,0 +1,11 @@
+.security-2fa-initial-setup__label span {
+    display: inline;
```
Spans are inline by default, so they shouldn't need us to specifically declare `display: inline`.

tyxla commented on client/me/security-2fa-initial-setup/style.scss:
```diff
@@ -0,0 +1,11 @@
+.security-2fa-initial-setup__label span {
+    display: inline;
+    margin: 0 0 0 5px;
+    padding: 0;
```
Is there any reason for resetting the padding? I don't see any padding there.

tyxla commented on client/me/security-2fa-initial-setup/style.scss:
```diff
@@ -0,0 +1,11 @@
+.security-2fa-initial-setup__label span {
+    display: inline;
+    margin: 0 0 0 5px;
```
Or we might as well just do `margin-left: 5px` since there is no other margin specified for this span.

tyxla commented on client/me/security-2fa-initial-setup/style.scss:
```diff
@@ -0,0 +1,11 @@
+    display: inline;
+    margin: 0 0 0 5px;
+    padding: 0;
+    font-size: 1rem; /* stylelint-disable-line */
```
Or we could just use `$font-body` instead of `1rem`? That way we wouldn't need to disable the linter.

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,12 +54,39 @@ class Security2faInitialSetup extends Component {
+					</FormLabel>
+				</div>
+
+				<FormButton onClick={ this.handleClick.bind( this ) }>
```
Instead of binding to `this`, we can just declare `handleClick` as an arrow function above. That will automatically bind it to `this`.

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -1,17 +1,27 @@
 
 	static propTypes = {
 		onSuccess: PropTypes.func.isRequired,
+		currentAuthMethod: PropTypes.func.isRequired,
```
I assume you meant to make this `string`?

```suggestion
		currentAuthMethod: PropTypes.string.isRequired,
```

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,12 +54,39 @@ class Security2faInitialSetup extends Component {
+						/>
+						<Gridicon icon="phone" />
+						<span>Set up using an app</span>
+						<p>
```
Noting that a `<p />` inside a label isn't valid HTML. Generally speaking, block elements can't be nested inside inline elements. Perhaps we can move the `<p />` outside of the label?

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,12 +54,39 @@ class Security2faInitialSetup extends Component {
+						<Gridicon icon="chat" />
+						<span>Set up using SMS</span>
+						<p>
+							WordPress.com will send you a SMS with a two-factor authentication codes when you log
```
Noting that we use both `two-factor` and `two-step` in the titles / descriptions. That can be confusing for the user - should we unify the wording?

xknown replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/57434/commits/b6b41aa21de25e45dca661a248fe428a0e57e902

tyxla commented on client/me/security-2fa-setup/index.jsx:
```diff
@@ -81,7 +85,10 @@ class Security2faSetup extends Component {
 				{ 'backup-codes' === this.state.step ? (
-					<Security2faSetupBackupCodes onFinished={ this.onFinished } />
+					<Security2faSetupBackupCodes
+						doSMSFlow={ this.state.authMethod !== 'app-based' }
```
Also, in case we add a third means of 2FA, should we be more precise here and instead of checking if we're not using app-based, check if we're on the SMS flow?

xknown replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/57434/commits/29c2e206070a1c7bd60b28cbed57278c8db9df23

tyxla reviewed (commented):
This appears to work great in my testing - nice work! 🙌 

I've left some feedback, mostly minor suggestions. Let me know if you have any questions. 

I'm happy to take another look - just re-request a review when you're ready 😉

xknown committed:
Use the appropriate type for currentAuthMethod

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Destructure this.props to simplify the code

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Use `isSmsFlow` instead of `doSMSFlow`

xknown committed:
Use a more explicit check for the SMS flow

This is to avoid potential issues if we decide to add another 2fa
method.

xknown committed:
Fix the initial setup flow

Don't longer use the currentAuthMethod prop because it doesn't work that
well when one clicks the Cancel button during the 2fa setup flow.

xknown committed:
Use an arrow function instead of class method.

xknown committed:
Use two-step instead of two-factor.

The former is used everywhere else.

xknown committed:
Fix typo

xknown committed:
Update the description text on the SMS flow.

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -23,6 +32,7 @@ class Security2faInitialSetup extends Component {
 	}
 
 	render() {
+		const setAuth = ( event ) => this.setState( { authMethod: event.currentTarget.value } );
```
Oops, this was likely meant to be declared as a separate method, just like `render()` is a separate method in the `Security2faInitialSetup` component class. Let's keep it an arrow function, though, in order to automatically bind to `this`.

xknown replied:
Fixed this in https://github.com/Automattic/wp-calypso/pull/57434/commits/698c5f29366127bf177213e94934e3c768857a35

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,10 +45,44 @@ class Security2faInitialSetup extends Component {
+						<FormRadio
+							name="auth_method"
+							value="app-based"
+							defaultChecked={ true }
```
Could be shortened to:

```suggestion
							defaultChecked
```

xknown commented:
@Automattic/editorial can I please get a review on the strings used in the UI. They are used along with the `this.props.translate( ... )` function.

tyxla commented on client/me/security-2fa-enable/index.jsx:
```diff
@@ -21,19 +21,19 @@ class Security2faEnable extends Component {
 
 	static defaultProps = {
-		doSMSFlow: false,
+		isSmsFlow: false,
```
Thanks for renaming the prop all the way 🙇‍♂️

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,10 +45,44 @@ class Security2faInitialSetup extends Component {
+						<span className="security-2fa-initial-setup__item-title">
+							{ this.props.translate( 'Set up using an app' ) }
+						</span>
+						<span className="security-2fa-initial-setup__description">
```
FWIW there's a `FormSettingExplanation` component in case that can do the job for you. Just note that it usually should be placed outside of the label, because it produces a block `<p>` tag. IMHO that's fine - I wouldn't expect the description to be part of the label.

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,10 +45,44 @@ class Security2faInitialSetup extends Component {
+							name="auth_method"
+							value="app-based"
+							defaultChecked={ true }
+							onChange={ setAuth }
```
Note that once we move `setAuth` to be a method of the component class, this will have to be changed to `this.setAuth`.

tyxla commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,10 +45,44 @@ class Security2faInitialSetup extends Component {
 					) }
 				</p>
 
+				<p>
```
I'd suggest wrapping both labels in a `<FormFieldset />` component. And do we really need the `<p>` tags now?

tyxla commented on client/me/security-2fa-sms-settings/index.jsx:
```diff
@@ -177,27 +174,15 @@ class Security2faSMSSettings extends Component {
-						<FormButton
-							disabled={ this.getSubmitDisabled() }
-							isPrimary={ false }
+							isPrimary={ true }
```
Can be shortened to:

```suggestion
							isPrimary
```

tyxla reviewed (commented):
I wanted to point out that if we pick "Set up using SMS" and proceed to the next step, then go back, "Set up using an app" is selected. Is that expected behavior? 

I've left some minor suggestions in the meantime while we're approving the copy. 

Thanks for the great work here 🙌

xknown committed:
Move setAuth out of the render method.

xknown committed:
Update client/me/security-2fa-sms-settings/index.jsx

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Update client/me/security-2fa-initial-setup/index.jsx

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

xknown committed:
Use the appropriate Form components.

xknown committed:
Remove unused css class

xknown committed:
Add a context for the new strings to be translated.

a8ci18n commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,10 +48,54 @@ class Security2faInitialSetup extends Component {
+					<FormLabel htmlFor="app-auth">
+						<Gridicon icon="phone" />
+						<span className="security-2fa-initial-setup__item-title">
+							{ this.props.translate( 'Set up using an app', {
```
:information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/me/security-2fa-initial-setup/index.jsx:
```diff
@@ -35,10 +48,54 @@ class Security2faInitialSetup extends Component {
+					<FormLabel htmlFor="sms-auth">
+						<Gridicon icon="chat" />
+						<span className="security-2fa-initial-setup__item-title">
+							{ this.props.translate( 'Set up using SMS', {
```
:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

xknown replied:
Resolved in fbdbdfa16ab525cc6a4f90470594869c5c7ed5bd

xknown committed:
Use comment instead of context

carolynannewells commented:
Hello! Carolyn here from Editorial. I have reviewed your new copy, and have a couple of suggestions, that I have included below: 

>Scan this QR code with your mobile app. 

_Scan this QR code with your authenticator app._ 

>First, we need your Mobile Phone number to send you verification codes when you choose the SMS method or in cases where the authenticator app on your phone is unavailable. 

_Please provide your mobile phone number. This enables us to send you verification codes via the SMS authentication method, or when the authenticator app on your phone is unavailable._ 

>Backup codes let you access your account if your phone is lost, stolen, or if you run it through the washing machine and the bag of rice trick does not work.

_Backup codes let you access your account if your phone is lost or stolen, or even just accidentally run through the washing machine!_

xknown committed:
Improve the suggested strings from the Editorial team

a8ci18n commented on client/me/security-2fa-backup-codes/index.jsx:
```diff
@@ -111,9 +111,7 @@ class Security2faBackupCodes extends Component {
-						'Backup codes let you access your account if your phone is ' +
-							'lost, stolen, or if you run it through the washing ' +
-							"machine and the bag of rice trick doesn't work."
+						'Backup codes let you access your account if your phone is lost or stolen, or even just accidentally run through the washing machine!'
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **31** times:
`translate( 'Backup codes let you access your account if your phone is lost, stolen, or if you run it through the washing machine and the bag of rice trick doesn't work.' )` *ES Score: 7*

a8ci18n commented on client/me/security-2fa-enable/index.jsx:
```diff
@@ -189,7 +189,7 @@ class Security2faEnable extends Component {
 				<p className="security-2fa-enable__qr-instruction">
 					{ this.props.translate(
-						"Scan this QR code with your mobile app. {{toggleMethodLink}}Can't scan the code?{{/toggleMethodLink}}",
+						"Scan this QR code with the authenticator app on your mobile. {{toggleMethodLink}}Can't scan the code?{{/toggleMethodLink}}",
```
:warning: This change will be queued for retranslation. We'll display the English text until then.
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **24** times:
`translate( 'Enter this time code into your mobile app. {{toggleMethodLink}}Prefer to scan the code?{{/toggleMethodLink}}' )` *ES Score: 7*

tyxla reviewed (approved):
LGTM 👍  Nice work!

One little note I wanted to mention in case it got lost: if we pick "Set up using SMS" and proceed to the next step, then go back, "Set up using an app" is selected. We might want to improve that in a follow-up, but it's not a blocker by any means.

xknown commented:
> One little note I wanted to mention in case it got lost: if we pick "Set up using SMS" and proceed to the next step, then go back, "Set up using an app" is selected. We might want to improve that in a follow-up, but it's not a blocker by any means.

I agree, added to my todo list so it doesn't get lost! :)

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6940403

Thank you @xknown for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6940403).
